### PR TITLE
[asan][Windows] Synchronizing ASAN init on Windows

### PR DESCRIPTION
**Pull request by zacklj89**
This is an attempt to synchronize ASAN initialization using `__sanitizer::atomic`s on Windows. One case in particular is a C# application that loads an ASan instrumented library.

This addresses:

1. [This check failure](https://github.com/llvm/llvm-project/blob/1df5ea29b43690b6622db2cad7b745607ca4de6a/compiler-rt/lib/asan/asan_interceptors.h#L29C36-L29C36) if a function is intercepted on T1 and called on T2 before T1 is done with ASAN initialization.
2. [This check failure](https://github.com/llvm/llvm-project/blob/1df5ea29b43690b6622db2cad7b745607ca4de6a/compiler-rt/lib/asan/asan_rtl.cpp#L387) can happen in parallel when two threads call `ENSURE_ASAN_INITED`.
3. [This check failure](https://github.com/llvm/llvm-project/blob/1df5ea29b43690b6622db2cad7b745607ca4de6a/compiler-rt/lib/asan/asan_rtl.cpp#L481) when thread is created before the main thread gets a chance to get created due to when `CreateThread` is intercepted.

Currently, everything is blocked by `#if SANITIZER_WINDOWS` because I'm not 100% sure about how ASan is loaded on other platforms and the implications.

SPR PRs [one](https://github.com/llvm/llvm-project/pull/74084) and [two](https://github.com/llvm/llvm-project/pull/74085)

### CODE REVIEW

File: compiler-rt/lib/asan/asan_thread.cpp [hunk 1]
```diff
@@ -93,6 +97,12 @@ AsanThreadContext *GetThreadContextByTidLocked(u32 tid) {
 AsanThread *AsanThread::Create(const void *start_data, uptr data_size,
                                u32 parent_tid, StackTrace *stack,
                                bool detached) {
+#if SANITIZER_WINDOWS
```
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_thread.cpp:100]: This block can be moved before `AsanThread *t = AsanThread::Create(params, current_tid, &stack, detached);` in `INTERCEPTOR_WINAPI(HANDLE, CreateThread`

File: compiler-rt/lib/asan/asan_thread.cpp [hunk 2]
```diff
@@ -288,11 +298,25 @@ void AsanThread::ThreadStart(tid_t os_id) {
 }
 
 AsanThread *CreateMainThread() {
```
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_thread.cpp:300]:
```
#ifdef // !SANITIZER_WINDOWS
AsanThread *CreateMainThread() {
...
#endif
```

And put windows version into asan_win.cpp

File: compiler-rt/lib/asan/asan_rtl.cpp [hunk 3]
```diff
@@ -71,16 +71,54 @@ static void CheckUnwind() {
...
-bool AsanInitIsRunning() { return asan_init_is_running == 1; }
+bool AsanInitIsRunning() {
+#if SANITIZER_WINDOWS
+  return atomic_load(&asan_init_is_running, memory_order_acquire) == 1;
+#else
+  return asan_init_is_running == 1;
+#endif
+}
+
+void CheckAsanInitRunning() {
```
Comment (vitalybuka) [inline compiler-rt/lib/asan/asan_rtl.cpp:114]: Probably not needed If you rebase on https://github.com/llvm/llvm-project/pull/74387

**Comment by vitalybuka** [comment]
Could you please if https://github.com/llvm/llvm-project/pull/74387 works for you?

**Comment by zacklj89** [comment]
I cherry-picked the change from #74387 for testing (alongside the change here in `asan_thread.cpp`). It appears sufficient after testing. Thanks!

**Pull request closed without merging**